Feature/optimization discrete variables

HugoKulesza opened the pull request.

flomnes committed:
Fix xpress identifier, use MIP solver if the problem is MIP

HugoKulesza committed:
glop solver activated
intVariable array added to store information on discrete variables

HugoKulesza committed:
Adding discrete value array in problem creation
Renaming intVar into VariablesEntieres
Allocation + filling values during optimization variables creation

HugoKulesza committed:
Giving discrete variables info to solvers

HugoKulesza committed:
Making use of MakeVar(min, max, isInteger, name)

HugoKulesza committed:
filling VariablesEntieres complete

HugoKulesza committed:
Deactivating heuristic and second optimization
(by commenting)
Only bringing back reduced costs and dual values when using MIP model

HugoKulesza committed:
Remove commented second optimization
ERRATUM for preceding commit :
reduced costs and dual values when using *LP* model

HugoKulesza committed:
Reactivated heuristics / second optimization
Adding an epsilon to int vars to measure stability through heuristics

HugoKulesza committed:
Use of round function for int vars out of XPRESS

HugoKulesza committed:
Test : remove epsilon to output of XPRESS
Check influence on heuristic

HugoKulesza committed:
Going back without 2nd optimization
Enabling solver output log

HugoKulesza committed:
remove commented code

HugoKulesza committed:
Renaming ucHeuristic and ucMILP
(resp. to ucFast and ucAccurate)

HugoKulesza committed:
also renaming selection methods

HugoKulesza committed:
Adding "milp" parameter as option of optimisation
conditionning "VariablesEntieres" to the use of milp parameter

HugoKulesza committed:
activating 2nd optimization (if not milp)

HugoKulesza committed:
No NODU smoothing needed for MILP resolution

HugoKulesza committed:
changed memory required for input with MILP

HugoKulesza committed:
(first draft) check of ortools if milp model

HugoKulesza committed:
error messages changes

flomnes committed:
Merge develop into feature/optimization_discrete_variables

HugoKulesza committed:
refactoring yearEndBuildFromThermalClusterIndex
separating smoothing and costs computing
not activating smoothing when MILP resolution

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/optimization_discrete_variables

flomnes commented on src/solver/optimisation/opt_construction_variables_couts_demarrages.cpp:
```diff
@@ -136,18 +144,21 @@ void OPT_ConstruireLaListeDesVariablesOptimiseesDuProblemeLineaireCoutsDeDemarra
               = NombreDeVariables;
             ProblemeAResoudre->TypeDeVariable[NombreDeVariables] = VARIABLE_NON_BORNEE;
+            ProblemeAResoudre->VariablesEntieres[NombreDeVariables] = false;
             NombreDeVariables++;
```
Not needed

flomnes commented on src/solver/utils/ortools_utils.cpp:
```diff
@@ -145,20 +160,38 @@ static void extract_from_MPSolver(const MPSolver* solver,
+    const bool isMIP = std::any_of(problemeSimplexe->VariablesEntieres.cbegin(),
+                                   problemeSimplexe->VariablesEntieres.cend(),
+                                   [](bool x) { return x; });
+
```
Duplicate code : add a const method in `PROBLEME_SIMPLEXE_NOMME`.

HugoKulesza committed:
reactivating NODU smoothing for MILP solve

HugoKulesza committed:
Merge branch 'feature/optimization_discrete_variables' of https://github.com/AntaresSimulatorTeam/Antares_Simulator into feature/optimization_discrete_variables

HugoKulesza committed:
fix case

flomnes commented on src/libs/antares/study/fwd.cpp:
```diff
@@ -293,8 +293,10 @@ UnitCommitmentMode StringToUnitCommitmentMode(const AnyString& text)
-    if (s == "accurate") // mixed integer linear problem
+        return ucFast;
+    if (s == "accurate") 
+        return ucAccurate;
```
Use `ucHeuristicFast`, `ucHeuristicAccurate`. When we come back to the code in a few months, we'll understand easily that "Fast" and "Accurate" both use heuristics, whereas MILP doesn't.

flomnes replied:
You may use `sed`, `xarg` and `find` to search & replace across the whole codebase. Something like
```
find -type f \( -name "*.cpp" -or -name "*.h" \) | xargs sed -i "s/search/replace/g"
```

flomnes commented on src/solver/application.cpp:
```diff
@@ -42,6 +42,26 @@ void printVersion()
+        {
+            throw Error::IncompatibleMILPOrtoolsSolver();
+        }
+    }
```
```suggestion
                       )
{
    using namespace Antares::Data;
    if (ucMode == UnitCommitmentMode::ucMILP)
    {
        if (!ortoolsUsed)
        {
            throw Error::IncompatibleMILPWithoutOrtools();
        }
        else if (ortoolsSolverUsed == OrtoolsSolver::sirius)
        {
            throw Error::IncompatibleMILPOrtoolsSolver();
        }
    }
```

flomnes commented on src/solver/application.cpp:
```diff
@@ -55,12 +75,13 @@ void checkSimplexRangeHydroPricing(Antares::Data::SimplexOptimization optRange,
     if (optRange == Antares::Data::SimplexOptimization::sorDay
-        && ucMode == Antares::Data::UnitCommitmentMode::ucMILP)
+        && (ucMode == Antares::Data::UnitCommitmentMode::ucAccurate
+            || ucMode == Antares::Data::UnitCommitmentMode::ucMILP))
```
```suggestion
        && (ucMode == UnitCommitmentMode::ucAccurate
            || ucMode == UnitCommitmentMode::ucMILP))
```
Add `using namespace Antares::Data;` as the 1st instruction in the function.

flomnes commented on src/solver/variable/state.cpp:
```diff
@@ -254,263 +255,285 @@ void State::yearEndBuildFromThermalClusterIndex(const uint clusterAreaWideIndex,
+                nivmin = ON_min[h];
 
-                    (ON_max[h] < ON_opt[h - 1]) ? nivmax = ON_max[h] : nivmax = ON_opt[h - 1];
+                (ON_max[h] < thermalClusterDispatchedUnitsCountForYear[h - 1]) 
+                        ? nivmax = ON_max[h] 
+                        : nivmax = thermalClusterDispatchedUnitsCountForYear[h - 1];
```
```suggestion
                 ON_max[h] = Math::Min<uint>(ON_max[h], thermalClusterDispatchedUnitsCountForYear[h - 1]);
```

flomnes commented on src/solver/variable/state.cpp:
```diff
@@ -254,263 +255,285 @@ void State::yearEndBuildFromThermalClusterIndex(const uint clusterAreaWideIndex,
-                thermalClusterProduction
-                  = thermalClusterAvailableProduction; // in mustrun, production==available
-                                                       // production
-            }
-            else
+            case Antares::Data::UnitCommitmentMode::ucFast:
```
This `switch..case` is used to determine `ON_min[h]`. It would be nice to make it visible here with a function
```cpp
ON_min[h] = minimalClustersON(unitCommitmentMode, currentCluster, /* ... */);
```

Lines 334-370.

flomnes commented on src/solver/variable/state.cpp:
```diff
@@ -254,263 +255,285 @@ void State::yearEndBuildFromThermalClusterIndex(const uint clusterAreaWideIndex,
+
+            ON_max[h] = static_cast<uint>(Math::Ceil(
+            thermalClusterAvailableProduction / currentCluster->nominalCapacityWithSpinning));
 
-                if (ON_max[h] < ON_min[h])
-                    ON_max[h] = ON_min[h];
+            if (currentCluster->minStablePower > 0.)
+            {
+                maxUnitNeeded = static_cast<uint>(
+                Math::Floor(thermalClusterProduction / currentCluster->minStablePower));
+                if (ON_max[h] > maxUnitNeeded)
+                    ON_max[h] = maxUnitNeeded;
             }
+
+            if (ON_max[h] < ON_min[h])
+                ON_max[h] = ON_min[h];
```
Same remark as above
```cpp
ON_max[h] = maximalClustersON(currentCluster, /* ... */);
```
If you can think of a better name, please use it.

HugoKulesza committed:
Cleaning, renaming after comments
Moving isMIP to named_problem

HugoKulesza committed:
adding methods to compute ON_Min and ON_Max

HugoKulesza committed:
remove unused variable

flomnes committed:
Merge branch develop into feature/optimization_discrete_variables

flomnes committed:
Apply clang-format

HugoKulesza committed:
remaking isMIP into a method

HugoKulesza committed:
removing const typing

HugoKulesza committed:
[TEMP] Test log xpress + 1 thread

HugoKulesza committed:
remove xpress parameters

HugoKulesza committed:
adding parameters
--ortools-parameters="" for specific solver parameters string
--ortools-verbose to activate verbosity

HugoKulesza committed:
[TEMP] removing ortools presolve

HugoKulesza committed:
Revert "[TEMP] removing ortools presolve"

This reverts commit 9500cecaecce71451eb41cba86bf9a9f57291ce7.

HugoKulesza committed:
Merge branch 'develop' into feature/optimization_discrete_variables

HugoKulesza committed:
Merge branch 'develop' into feature/optimization_discrete_variables

HugoKulesza committed:
changing parameters setting for ortools solvers

flomnes commented on src/solver/utils/ortools_utils.cpp:
```diff
@@ -143,7 +143,10 @@ MPSolver* convert_to_MPSolver(
     MPSolver* solver = new MPSolver("simple_lp_program", solverType);
 
-    tuneSolverSpecificOptions(solver);
+    if (study.parameters.ortoolsParamsString == "")
```
```suggestion
    if (study.parameters.ortoolsParamsString.empty())
```

HugoKulesza committed:
2 solvers kept in memory for optim 1/2
for milp, 2nd optim with min/max fixed for M

flomnes commented on src/solver/optimisation/opt_alloc_probleme_a_optimiser.cpp:
```diff
@@ -162,10 +162,15 @@ static void optimisationAllocateProblem(PROBLEME_HEBDO* ProblemeHebdo, const int
 
-    ProblemeAResoudre->ProblemesSpx = (PROBLEMES_SIMPLEXE*)MemAlloc(sizeof(PROBLEMES_SIMPLEXE));
-    ProblemeAResoudre->ProblemesSpx->ProblemeSpx = (void**)MemAlloc(NbIntervalles * sizeof(void*));
+    ProblemeAResoudre->ProblemesSpx1 = (PROBLEMES_SIMPLEXE*)MemAlloc(sizeof(PROBLEMES_SIMPLEXE));
+    ProblemeAResoudre->ProblemesSpx1->ProblemeSpx = (void**)MemAlloc(NbIntervalles * sizeof(void*));
```
Simplification
```cpp
...->ProblemeSpx[OptIndex]->ProblemeSpx = ...
```

flomnes commented on src/solver/optimisation/opt_appel_solveur_lineaire.cpp:
```diff
@@ -103,15 +103,21 @@ bool OPT_AppelDuSimplexe(PROBLEME_HEBDO* ProblemeHebdo, uint numSpace, int NumIn
-    ProbSpx = (PROBLEME_SPX*)(ProblemeAResoudre->ProblemesSpx->ProblemeSpx[(int)NumIntervalle]);
-    solver = (MPSolver*)(ProblemeAResoudre->ProblemesSpx->ProblemeSpx[(int)NumIntervalle]);
+    PROBLEMES_SIMPLEXE* ProblemesSpx;
+    if (Probleme.NumeroOptimisation == 1)
+        ProblemesSpx = ProblemeAResoudre->ProblemesSpx1;
+    else if (Probleme.NumeroOptimisation == 2)
+        ProblemesSpx = ProblemeAResoudre->ProblemesSpx2;
```
L'utilisation d'un index d'optimisation (0 ou 1) permet d'éliminer le if...else

flomnes commented on src/solver/optimisation/opt_nombre_min_groupes_demarres_couts_demarrage.cpp:
```diff
@@ -206,23 +206,32 @@ void OPT_AjusterLeNombreMinDeGroupesDemarresCoutsDeDemarrage(PROBLEME_HEBDO* Pro
+                              * NombreMaxDeGroupesEnMarcheDuPalierThermique[PdtHebdo]
+                            > PuissanceDisponibleDuPalierThermique[PdtHebdo])
+                            PuissanceDisponibleDuPalierThermique[PdtHebdo]
+                              = PminDUnGroupeDuPalierThermique
+                                * NombreMaxDeGroupesEnMarcheDuPalierThermique[PdtHebdo];
+                    }
+                }
+                else
                 {
-                    if (NombreMaxDeGroupesEnMarcheDuPalierThermique[PdtHebdo]
-                        < NombreMinDeGroupesEnMarcheDuPalierThermique[PdtHebdo])
-                        NombreMaxDeGroupesEnMarcheDuPalierThermique[PdtHebdo]
-                          = NombreMinDeGroupesEnMarcheDuPalierThermique[PdtHebdo];
-
-                    if (PminDUnGroupeDuPalierThermique
-                          * NombreMaxDeGroupesEnMarcheDuPalierThermique[PdtHebdo]
-                        > PuissanceDisponibleDuPalierThermique[PdtHebdo])
-                        PuissanceDisponibleDuPalierThermique[PdtHebdo]
-                          = PminDUnGroupeDuPalierThermique
-                            * NombreMaxDeGroupesEnMarcheDuPalierThermique[PdtHebdo];
+                    NombreMaxDeGroupesEnMarcheDuPalierThermique[PdtHebdo]
+                      = NombreMinDeGroupesEnMarcheDuPalierThermique[PdtHebdo];
```
Ça me paraît violent comme changement, on en discute

flomnes commented on src/solver/utils/ortools_utils.cpp:
```diff
@@ -165,21 +195,36 @@ static void extract_from_MPSolver(const MPSolver* solver,
     {
         auto& row = constraints[idxRow];
-        problemeSimplexe->CoutsMarginauxDesContraintes[idxRow] = row->dual_value();
+        if (isMIP && problemeSimplexe->NumeroOptimisation == 1)
+        {
+            problemeSimplexe->CoutsMarginauxDesContraintes[idxRow] = 0;
+        }
+        else
+        {
+            problemeSimplexe->CoutsMarginauxDesContraintes[idxRow] = row->dual_value();
+        }
```
Si MIP, il faut lire la valeur de la 2nde optimisation, qui est un LP résolu exprès pour avoir les valeurs duales

HugoKulesza committed:
[TEMP] individual year objectives

HugoKulesza committed:
Revert "[TEMP] individual year objectives"

This reverts commit 7d6164c2683e4514317f4f15ac17e76f4a7d0643.

HugoKulesza committed:
fix : fixing NODU min and max for ALL timesteps
in MILP mode

guilpier-code commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -318,6 +318,8 @@ void Parameters::reset()
 
     ortoolsUsed = false;
     ortoolsEnumUsed = OrtoolsSolver::sirius;
+    ortoolsParamsString = "";
```
`ortoolsParamsString.clear()` would be better

guilpier-code commented on src/libs/antares/study/parameters.h:
```diff
@@ -552,8 +552,12 @@ class Parameters final
+    //! Ortools solver used for simulation
     OrtoolsSolver ortoolsEnumUsed;
+    //! Specific parameters string for ortools solver
+    string ortoolsParamsString;
```
`string ortoolsParamsString;` : 
What this parameter is for exactly ? It's not obvious when reading its name.
Maybe it means something to someone that knows about or-tools, but not to me. 
We already know it's a string, so we don't need it in the name, and **Params** in the name does not give us extra clue, so it's not needed as well.

guilpier-code replied:
`string ortoolsParamsString;` (again) :
If this parameter can have several possible values, making it an **enum** would be better.
**To be noticed** : there is no check of any kind on its value, unless this check is done inside or-tools. Theoretically, this parameter's value could be any string (most checks are done in class **Parameters**).

flomnes replied:
Removed

guilpier-code commented on src/solver/utils/ortools_utils.cpp:
```diff
@@ -126,21 +133,39 @@ MPSolver* convert_to_MPSolver(
     auto& study = *Data::Study::Current::Get();
```
`MPSolver* convert_to_MPSolver(const PROBLEME_SIMPLEXE_NOMME* problemeSimplexe)` : 
Like many functions/methods in Antares, there is a dependency problem with function : it depends on the whole study, while it should take a few values as argument (for instance **ortoolsEnumUsed**, or the new **ortoolsParamsString**).
Passing these parameters would break the dependency to study, and as a side effect, not being recompiled when we touch the Study

flomnes replied:
Removed, the `Study` singleton has been removed.

guilpier-code commented on src/solver/utils/ortools_utils.cpp:
```diff
@@ -126,21 +133,39 @@ MPSolver* convert_to_MPSolver(
-    MPSolver::OptimizationProblemType solverType
-      = OrtoolsUtils().getLinearOptimProblemType(study.parameters.ortoolsEnumUsed);
+
+    MPSolver::OptimizationProblemType solverType;
```
In this file, we have 3 **if** statements following the pattern : 
```C++
if (isMIP && problemeSimplexe->NumeroOptimisation == 1)
{
     ....
}
else
{
     ....
}
``` 
This is the sign that we should use polymorphism and encapsulate **MIP** problem and **LP** problem in separate classes.
Let's create an interface (or a base class to factorize the common code of both problems), and have a factory to instantiate the right problem in a base pointer, etc.

If we do that (and if don't want to touch code but in the current source file), we will be lead to  : 
- change the entry point, that is the **ORTOOLS_Simplexe(...)** function's body : call the factory and the resolution there
- extract the **solver type** determination (sirius, glop, xpress, ...) from **convert_to_MPSolver(...)** and put it in the factory
- make several functions of this file methods of LP / MIP classes

flomnes replied:
This bit of code has been removed

guilpier-code commented on src/solver/utils/ortools_utils.cpp:
```diff
@@ -16,11 +16,15 @@ static void transferVariables(MPSolver* solver,
-                              const std::vector<std::string>& NomDesVariables)
-{
+                              const std::vector<std::string>& NomDesVariables,
+                              const std::vector<bool>& VariablesEntieres,
```
If we refactor as suggested below, **VariablesEntieres** will be used only in the **MIP** problem

guilpier-code commented on src/solver/simulation/solver.hxx:
```diff
@@ -852,7 +852,8 @@ void ISimulation<Impl>::estimateMemoryForOptimizationPb(Antares::Data::StudyMemo
     }
 
     // Mode accurate
```
`// Mode accurate` ==> `// Mode accurate and MILP`

guilpier-code commented on src/solver/application.cpp:
```diff
@@ -286,6 +307,9 @@ void Application::prepare(int argc, char* argv[])
 
     // Some more checks require the existence of pParameters, hence of a study.
     // Their execution is delayed up to this point.
+    checkOrtoolsUsage(
```
Does the comment : 
```C++    
// Some more checks require the existence of pParameters, hence of a study.
// Their execution is delayed up to this point.
```
regard also the **checkOrtoolsUsage**'s call ?
I ask it because the comment used to regard the following call.
If the comment regards **checkOrtoolsUsage**'s call as well, where are the other checks ? And couldn't we gather all checks in one function, called at the right place ?

guilpier-code commented on src/ui/simulator/windows/options/advanced/advanced.cpp:
```diff
@@ -841,41 +841,65 @@ void AdvancedParameters::onUnitCommitmentMode(Component::Button&, wxMenu& menu,
 }
 
-void AdvancedParameters::onSelectUCHeuristic(wxCommandEvent& /* evt */)
+void AdvancedParameters::onSelectUCHeuristicFast(wxCommandEvent& /* evt */)
```
the 3 callback functions associated to each unit commitment type are very similar and differ by this type only. We should avoid code duplication by creating a template function.

flomnes replied:
Yes, agreed

guilpier-code commented on src/ui/simulator/windows/options/advanced/advanced.cpp:
```diff
@@ -841,41 +841,65 @@ void AdvancedParameters::onUnitCommitmentMode(Component::Button&, wxMenu& menu,
     wxString text;
 
-    text = wxStringFromUTF8(UnitCommitmentModeToCString(Data::ucHeuristic)); // Fast
+    text = wxStringFromUTF8(UnitCommitmentModeToCString(Data::ucHeuristicFast)); // Fast
```
There is code duplication when creating unit commitment items. Each creation is similar to the other. Could we factorize that code in a single function ?

guilpier-code commented on src/solver/optimisation/opt_appel_solveur_lineaire.cpp:
```diff
@@ -103,15 +103,21 @@ bool OPT_AppelDuSimplexe(PROBLEME_HEBDO* ProblemeHebdo, uint numSpace, int NumIn
+        ProblemesSpx = ProblemeAResoudre->ProblemesSpx1;
+    else if (Probleme.NumeroOptimisation == 2)
+        ProblemesSpx = ProblemeAResoudre->ProblemesSpx2;
+
```
So many problems : 
- `PROBLEME_HEBDO* ProblemeHebdo`
- `PROBLEME_SIMPLEXE_NOMME Probleme`
- `PROBLEME_ANTARES_A_RESOUDRE* ProblemeAResoudre;`
- `PROBLEMES_SIMPLEXE* ProblemesSpx;`
-  `PROBLEME_SPX* ProbSpx`
-  `MPSolver* solver`
- `ProblemeAResoudre->ProblemesSpx1;`
- `ProblemeAResoudre->ProblemesSpx2;`

Do we really need them all ? 
It's very likely that there is unnecessary complexity here.
We **really** should do something to simplify this. It's always painful to read this area of code.

flomnes replied:
I agree ! It should be quite easy when/if we switch to OR-Tools only, discarding the "native/sirius" mode.

guilpier-code commented on src/solver/optimisation/opt_appel_solveur_lineaire.cpp:
```diff
@@ -103,15 +103,21 @@ bool OPT_AppelDuSimplexe(PROBLEME_HEBDO* ProblemeHebdo, uint numSpace, int NumIn
-    ProbSpx = (PROBLEME_SPX*)(ProblemeAResoudre->ProblemesSpx->ProblemeSpx[(int)NumIntervalle]);
-    solver = (MPSolver*)(ProblemeAResoudre->ProblemesSpx->ProblemeSpx[(int)NumIntervalle]);
+    PROBLEMES_SIMPLEXE* ProblemesSpx;
+    if (Probleme.NumeroOptimisation == 1)
```
**PREMIERE_OPTIMISATION** and **DEUXIEME_OPTIMISATION** pre-processor defined constants exist. Should not we use them ?

guilpier-code replied:
**Question** : does it mean we allocate 2 problems instead of one ? Is needed RAM growing ?

flomnes replied:
We should use the constants.

The 2 problems are required for the 2nd resolution, which is LP and allows us to recover dual values/reduced costs for continuous variables.

flomnes replied:
Although with a bit of work, we may re-use the existing problem for the 2 resolutions, changing only variable types

- Use the real variable type (continuous/integer)
- Recover values for integer values
- Fix integer variables to the above
- Solve as an LP problem, profit

flomnes replied:
Can't do this with OR-Tools, since MPSolver's LP/MIP status is immutable. Unfortunately, we'll need a copy of the MPSolver object.

guilpier-code commented on src/solver/optimisation/opt_structure_probleme_a_resoudre.h:
```diff
@@ -140,6 +141,9 @@ typedef struct
     std::vector<std::string> NomDesVariables;
     std::vector<std::string> NomDesContraintes;
+
+    std::vector<bool> VariablesEntieres; // true = int, false = continuous
```
There is already a std::vector\<bool\> **VariablesEntieres** as a member of class **PROBLEME_SIMPLEXE_NOMME**.
It looks like code duplication to me.

flomnes replied:
The `VariablesEntieres` defined in `PROBLEME_SIMPLEXE_NOMME` is a const reference to this one.

guilpier-code commented on src/solver/optimisation/opt_alloc_probleme_a_optimiser.cpp:
```diff
@@ -178,6 +183,7 @@ static void optimisationAllocateProblem(PROBLEME_HEBDO* ProblemeHebdo, const int
 
     ProblemeAResoudre->NomDesVariables.resize(ProblemeAResoudre->NombreDeVariables);
     ProblemeAResoudre->NomDesContraintes.resize(ProblemeAResoudre->NombreDeContraintes);
+    ProblemeAResoudre->VariablesEntieres.resize(ProblemeAResoudre->NombreDeVariables);
```
`ProblemeAResoudre->VariablesEntieres.resize(...)` : 
So here we reserve memory even if a **MILP** problem is not required.

guilpier-code commented on src/solver/optimisation/opt_construction_variables_couts_demarrages.cpp:
```diff
@@ -79,6 +79,8 @@ void OPT_ConstruireLaListeDesVariablesOptimiseesDuProblemeLineaireCoutsDeDemarra
                   = NombreDeVariables;
                 ProblemeAResoudre->TypeDeVariable[NombreDeVariables]
                   = VARIABLE_BORNEE_DES_DEUX_COTES;
+                if (ProblemeHebdo->OptimisationAvecVariablesEntieres)
```
This source file is originally about the kind of bounds we put on variables regarding the number of  working dispatchable units.
What we add here : whether the problem to solve is a **LP** or **MILP**, these variables are flagged as discrete or not.
So, on flagging variables like this, we : 
- mix problem **LP** and problem **MILP**, which contradict separation of concerns
- mix some operations that are not really of the same nature

We could create a new function that only gives a flag (discrete or not) variables related to number of  working dispatchable units, and call it as follows : 
```C++
if (ProblemeHebdo->OptimisationAvecVariablesEntieres)
    OPT_makeNODUvariablesDiscrete(ProblemeHebdo);
```

guilpier-code commented on src/solver/optimisation/opt_nombre_min_groupes_demarres_couts_demarrage.cpp:
```diff
@@ -206,23 +206,35 @@ void OPT_AjusterLeNombreMinDeGroupesDemarresCoutsDeDemarrage(PROBLEME_HEBDO* Pro
                     NombreMinDeGroupesEnMarcheDuPalierThermique[PdtHebdo] = (int)ceil(X);
                 }
 
```
We should think of whether we want to keep **ProblemeHebdo->Expansion**, and remove it otherwise.
Here we would remove dozens of lines

flomnes replied:
Xpansion sometimes uses it. And it is outside the scope of this PR

guilpier-code commented on src/solver/optimisation/opt_nombre_min_groupes_demarres_couts_demarrage.cpp:
```diff
@@ -206,23 +206,35 @@ void OPT_AjusterLeNombreMinDeGroupesDemarresCoutsDeDemarrage(PROBLEME_HEBDO* Pro
-                  ProblemeHebdo, NombreMinDeGroupesEnMarcheDuPalierThermique, Pays, Index);
-
-                for (PdtHebdo = 0; PdtHebdo < NombreDePasDeTempsProblemeHebdo; PdtHebdo++)
+                if (ProblemeHebdo->OptimisationAvecVariablesEntieres == NON_ANTARES)
```
Same remark as in **src/solver/optimisation/opt_construction_variables_couts_demarrages.cpp**, we should think of separating **LP** and **MILP** regarding fixing Min number of working dispatchable units

flomnes committed:
Merge remote-tracking branch 'github/develop'

flomnes committed:
Apply clang-format

flomnes committed:
Merge remote-tracking branch 'github/release/8.6.x' into HEAD

flomnes committed:
Fix build, formatting

flomnes committed:
Remove ortoolsVerbose & ortoolsParamsString

flomnes committed:
Simplify condition a bit

flomnes committed:
Fix variable type in MPS writer

flomnes commented on src/solver/utils/ortools_utils.cpp:
```diff
@@ -175,21 +164,36 @@ static void extract_from_MPSolver(const MPSolver* solver,
     {
         auto& row = constraints[idxRow];
-        problemeSimplexe->CoutsMarginauxDesContraintes[idxRow] = row->dual_value();
+        if (isMIP && problemeSimplexe->NumeroOptimisation == 1)
+        {
```
Use `PREMIERE_OPTIMISATION`, etc.

flomnes committed:
[skip ci] Add TODO

flomnes committed:
First try at merging

flomnes committed:
Add forgotten data member

flomnes committed:
Revert state.*

flomnes committed:
One more fix

flomnes committed:
Fix build for GUI

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/optimization_discrete_variables

flomnes committed:
Factorize UI code a bit

flomnes committed:
Move `checkOrtoolsUsage`

flomnes committed:
Add comment

flomnes committed:
Restore existing behavior for unitCommitmentMode = ucHeuristicAccurate

flomnes committed:
Small clean-up

flomnes committed:
Restore dual values & reduced costs

flomnes committed:
Renaming

flomnes committed:
Setup relaxation

flomnes committed:
Revert "Setup relaxation"

This reverts commit 1c6411b1969c7813d3d9b5f83efb00d9e135aa97.

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/optimization_discrete_variables

flomnes committed:
Remove relaxation, use 0 for reduced cost & marginal values

flomnes committed:
Format

flomnes commented on src/solver/optimisation/opt_alloc_probleme_a_optimiser.cpp:
```diff
@@ -81,9 +81,11 @@ void OPT_AllocateFromNumberOfVariableConstraints(PROBLEME_ANTARES_A_RESOUDRE* Pr
+    ProblemeAResoudre->NomDesVariables.resize(ProblemeAResoudre->NombreDeVariables);
+    ProblemeAResoudre->NomDesContraintes.resize(ProblemeAResoudre->NombreDeContraintes);
+    // Integer variables ? (MILP)
+    ProblemeAResoudre->VariablesEntieres.resize(ProblemeAResoudre->NombreDeVariables);
```
Partially revert to use `nbVariables` for consistency

flomnes committed:
Partial revert

flomnes committed:
Disable variables requiring dual values / marg costs when ucMode = milp

sylvlecl commented on src/libs/antares/study/parameters.h:
```diff
@@ -433,11 +432,13 @@ class Parameters final
         //! Unit Commitment Mode
         UnitCommitmentMode ucMode;
-    } unitCommitment;
+        void addExcludedVariables(std::vector<std::string>&) const;
```
Not obvious: we could add a comment why this is needed

flomnes replied:
done

sylvlecl commented on src/libs/antares/study/fwd.h:
```diff
@@ -464,8 +464,9 @@ SheddingPolicy StringToSheddingPolicy(const AnyString& text);
 {
-    ucHeuristic = 0,
-    ucMILP, // mixed integer linear problem
+    ucHeuristicFast = 0,
```
Would be nice to have some explanatory comments on that enum values too, I think.
The old `ucMILP` is the new `ucHeuristicAccurate`, that's right ?

flomnes replied:
yes, added some comments

The old ucMILP wasn't MILP

sylvlecl commented on src/solver/application.cpp:
```diff
@@ -31,6 +31,10 @@ using namespace Antares::Check;
 
 namespace
 {
+// CHECK nécessaire d'un choix simultané entre " unit-commitment = MILP " et " ortools-used "
```
I think you have handled that TODO

flomnes replied:
It is now handled, I will now remove this TODO note.
```
[2023-10-12 10:05:03][solver][fatal] Unit Commitment mode 'milp' must be used with an OR-Tools solver 
```

sylvlecl reviewed (commented):
Nice:
it would be great to have a small in memory end-to-end test as in `simple-study.cpp`, I think it's possible with one of the solvers available with ortools ?

flomnes committed:
Add comments

flomnes committed:
Add comments in `enum UnitCommitmentMode`

flomnes committed:
Remove obsolete TODO note

flomnes committed:
Add OutputRetriever::thermalNbUnitsON

flomnes committed:
Add 2 unit tests with COIN

flomnes committed:
Remove obsolete TODO note

flomnes committed:
Fix formatting

flomnes committed:
Fix formatting

flomnes committed:
Skip 2nd optimization when solving with integer variables

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/optimization_discrete_variables

flomnes committed:
[skip ci] Update ref guide
